lib/src/stream_queue.dart

```diff
@@ -407,6 +458,84 @@ class _StreamQueue<T> extends StreamQueue<T> {
   }
 }
 
+/// A [StreamQueue] cloned from a parent by [StreamQueue.startTransactions].
+///
+/// This queue may be [accept]ed, moving the parent to the same event as this,
+/// or [reject]ed, marking this as finished without modifying the parent.
+class TransactionStreamQueue<T> extends _StreamQueue<T> {
+  /// The request that created this queue.
+  final _TransactionRequest<T> _request;
+
+  /// Whether [accept] has been called.
+  var _accepted = false;
+
+  /// Whether [reject] has been called.
+  var _rejected = false;
+
+  /// The total number of events received by this queue, including events that
+  /// haven't yet been consumed by requests.
+  ///
+  /// This is used to fast-forward the parent queue if this transaction is
+  /// accepted.
+  var _eventsReceived = 0;
+
+  TransactionStreamQueue._(Stream<T> sourceStream, this._request)
+      : super(sourceStream);
+
+  /// Updates the parent [StreamQueue] to match this transaction's state.
+  ///
+  /// Further requests on this queue will be completed as though the underlying
+  /// stream had been closed, as though [cancel] were called with `immediate:
+  /// true`. Other transactions in the same batch are automatically rejected.
+  ///
+  /// Throws a [StateError] if [accept] or [reject] has already been called, or
+  /// if there are pending requests on this queue.
+  void accept() {
```

Comment:
> (Or is this prevented because the `startTransaction` request would make the request-queue non-empty?)

Yep. An open transaction counts as a pending request.